src/agent/src/main.rs

```diff
@@ -394,6 +417,103 @@ async fn start_sandbox(
     Ok(())
 }
 
+// Start-up attestation-agent, CDH and api-server-rest if they are packaged in the rootfs
+fn init_attestation_components(logger: &Logger, _config: &AgentConfig) -> Result<()> {
+    // The Attestation Agent will run for the duration of the guest.
+    launch_process(
+        logger,
+        AA_PATH,
+        &vec!["--attestation_sock", AA_ATTESTATION_URI],
+        AA_ATTESTATION_SOCKET,
+        DEFAULT_LAUNCH_PROCESS_TIMEOUT,
+    )
+    .map_err(|e| anyhow!("launch_process {} failed: {:?}", AA_PATH, e))?;
+
+    let config_path = OCICRYPT_CONFIG_PATH;
+
+    // The image will need to be encrypted using a keyprovider
+    // that has the same name (at least according to the config).
+    let ocicrypt_config = serde_json::json!({
+        "key-providers": {
+            "attestation-agent":{
+                "ttrpc":CDH_SOCKET
+            }
+        }
+    });
+
+    fs::write(config_path, ocicrypt_config.to_string().as_bytes())?;
+
+    if let Err(e) = launch_process(
+        logger,
+        CDH_PATH,
+        &vec![],
+        CDH_SOCKET,
+        DEFAULT_LAUNCH_PROCESS_TIMEOUT,
+    ) {
+        error!(logger, "launch_process {} failed: {:?}", CDH_PATH, e);
+    } else {
+        let features = _config.guest_components_rest_api;
+        match features {
+            GuestComponentsFeatures::None => {}
+            _ => {
+                if let Err(e) = launch_process(
+                    logger,
+                    API_SERVER_PATH,
+                    &vec!["--features", &features.to_string()],
+                    "",
+                    0,
+                ) {
+                    error!(logger, "launch_process {} failed: {:?}", API_SERVER_PATH, e);
```

Comment:
Should we raise an error for this function if ASR fails to run?

Reply:
I guess the original theory was that the CDH and ASR are more optional than the attestation-agent. Is your suggestion that we replace the logging of the error with returning the error?

Reply:
Here are some minimal runtime requirements.
|kata feature|AA|CDH|ASR|feature of  `image-rs`|
|---|---|---|---|---|
|guest-pull|N|N|N|`default`|
|decrypt-image|N(offline-fs-kbc), Y(cc-kbc) *|Y|N(restapi feature = none), Y(or else) **| `kata-cc-rustls-tls`|
|cosign-verification|N(offline-fs-kbc), Y(cc-kbc)|Y|N(restapi feature = none), Y(or else)| `kata-cc-rustls-tls`|
|auth|N(offline-fs-kbc), Y(cc-kbc)|Y|N(restapi feature = none), Y(or else)| `kata-cc-rustls-tls`|

As * and ** are determined at runtime, so we must support both during compilation time. Thus the table would be translated to the following compilation time requirements.

|kata feature|AA|CDH|ASR|feature of  `image-rs`|
|---|---|---|---|---|
|guest-pull|N|N|N|`default`|
|decrypt-image|Y|Y|Y| `kata-cc-rustls-tls`|
|cosign-verification|Y|Y|Y| `kata-cc-rustls-tls`|
|auth|Y|Y|Y| `kata-cc-rustls-tls`|

So, maybe we should detect the current kata-agent's feature to decide whether **all** or **none** of the three components to be launched?

Another way is to add `attestation` feature (whether we support `AA`, s.t. if `cc-kbc` is supported) and `asr` feature (whether we support `ASR`, s.t. if `restapi-feature` is supported) to control composition in a more granular way.

There must be some trade-offs.

Reply:
I think we are at risk of over optimising this and heading towards more service management features, that is probably part of a wider issue and discussion. If we move image pull into the CDH that will change this behaviour for guest-pull too, so maybe we leave it with the logging for now and re-visit this later as we have a clearer idea what the long term approach is?

Reply:
Thanks @Xynnn007 for the nice summary tables. And I agree with @stevenhorsman 's proposal.